### PR TITLE
Refactor metrics

### PR DESCRIPTION
**Pull request by SuperQ**
In order to make it easier to differentiate between `SHOW SLAVE STATUS`
and `SHOW GLOBAL STATUS` we need add a `Subsystem` to metric names.

Refactor `SHOW GLOBAL STATUS` pattern matching

This will make room in the metric namespace for MySQL metrics from the
`performance_schema` tables.

### CODE REVIEW

**Comment by SuperQ** [comment]
@fabxc @matthiasr This will break all the existing metric names, but is necessary to avoid collisions between GLOBAL and SLAVE status variables.

**Comment by fabxc** [comment]
:+1: To feel pretty one could move the source strings to constants.

**Comment by SuperQ** [comment]
@fabxc Added `slave_status` and `global_status` constants.

File: mysqld_exporter.go [hunk 1]
```diff
@@ -17,7 +17,9 @@ import (
 )
 
 const (
-	namespace = "mysql"
+	namespace     = "mysql"
+	slave_status  = "slave_status"
```
Comment (fabxc) [inline mysqld_exporter.go]: camelCase in Go.

**Comment by SuperQ** [comment]
@fabxc I agree on the common prefix approach in general, but in this case it would be confusing from the MySQL perspective.  For example the next PR will be called "performance" to represent the `performance_schema` subsystem.

**Comment by fabxc** [comment]
I see. LGTM :+1:

**Comment by SuperQ** [comment]
@beorn7 PTAL

**Comment by beorn7** [comment]
Looking...

File: mysqld_exporter.go [hunk 2]
```diff
@@ -225,36 +241,47 @@ func (e *Exporter) setGenericMetric(name string, value float64) {
...
+		e.commands.With(prometheus.Labels{"command": command[1]}).Set(value)
+	case len(connectionError) > 0:
+		e.connectionErrors.With(prometheus.Labels{"error": connectionError[1]}).Set(value)
+	case len(innodbRow) > 0:
+		e.innodbRows.With(prometheus.Labels{"operation": innodbRow[1]}).Set(value)
+	case len(performanceSchema) > 0:
+		e.performanceSchema.With(prometheus.Labels{"instrumentation": performanceSchema[1]}).Set(value)
+	default:
+		e.setGenericMetric(globalStatus, name, value)
+	}
```
Comment (beorn7) [inline mysqld_exporter.go]: Flow is a bit clunky here. Also, all regexp's are evaluation although we only need one hit.

How about the following:

if command := comRegex.FindStringSubmatch(name); command != nil {
    e.commands.With(prometheus.Labels{"command": command[1]}).Set(value)
    return
}
if //...
// and finally:
e.setGenericMetric(globalStatus, name, value)

Comment (SuperQ) [inline mysqld_exporter.go]: Yea, I didn't write this part.  I can refactor it as well.

**Comment by SuperQ** [comment]
@beorn7 Ok, I refactored the regexp, much better now.

File: mysqld_exporter.go [hunk 3]
```diff
@@ -225,36 +241,42 @@ func (e *Exporter) setGenericMetric(name string, value float64) {
 	e.metrics[name].Set(value)
 }
 
-func (e *Exporter) setMetrics(scrapes <-chan []string) {
-	var comRegex = regexp.MustCompile(`^com_(.*)$`)
-	var connectionErrorRegex = regexp.MustCompile(`^connection_errors_(.*)$`)
-	var innodbRowRegex = regexp.MustCompile(`^innodb_rows_(.*)$`)
-	var performanceSchemaRegex = regexp.MustCompile(`^performance_schema_(.*)$`)
+var (
+	globalStatusRegexp = regexp.MustCompile(`^(com|connection_errors|innodb_rows|performance_schema)_(.*)$`)
```
Comment (beorn7) [inline mysqld_exporter.go]: Since this is a single var now, you should just write

    var globalStatusRegexp = regexp...

**Comment by beorn7** [comment]
:+1: otherwise.

**Pull request merged**